include/mbedtls/rsa.h

```diff
@@ -1021,16 +1021,14 @@ int mbedtls_rsa_rsassa_pkcs1_v15_verify( mbedtls_rsa_context *ctx,
  * \brief          This function performs a PKCS#1 v2.1 PSS verification
  *                 operation (RSASSA-PSS-VERIFY).
  *
- *                 The hash function for the MGF mask generating function
- *                 is that specified in the RSA context.
- *
- * \note           The \p hash_id in the RSA context is the one used for the
- *                 verification. \p md_alg in the function call is the type of
- *                 hash that is verified. According to <em>RFC-3447: Public-Key
- *                 Cryptography Standards (PKCS) #1 v2.1: RSA Cryptography
- *                 Specifications</em> it is advised to keep both hashes the
- *                 same. If \p hash_id in the RSA context is unset,
- *                 the \p md_alg from the function call is used.
+ * \note           The \c hash_id in the \p ctx context selects the hash used
+ *                 for the encoding operation and for the mask generation
+ *                 function (MGF1). For more details on the encoding operation
+ *                 and the mask generation function, consult <em>RFC-3447:
+ *                 Public-Key Cryptography Standards (PKCS) #1 v2.1: RSA
+ *                 Cryptography Specifications</em>. If \c hash_id in the
+ *                 \p ctx context is #MBEDTLS_MD_NONE, the \p md_alg parameter
+ *                 is used.
```

Comment:
I agree, to both points, this has been bothering me for a long time too, especially because it allows non-compliant behaviour.

(Also, the documentation is still incorrect, I'll update that as well.)

Reply:
(Apparently the documentation problem was only present in a commit message, did a rebase to fix it.)

Reply:
> Usually it's the other way round. I'd like to error out instead if the two don't match, rather than do something surprising. This would apply restrictions discussed in #4561 to the lower level RSA module.
> 
> Likewise, instead of ignoring `hashlen`, I think the function should error out if it doesn't match. For `hashlen`, I think 0 should be rejected: the caller should always know how long the input buffer is.

@gilles-peskine-arm Is the recommendation to change the documentation/functionality of almost all the functions in `rsa.h` so that `hashlen` is no longer overwritten and instead is checked to match the expected length of `md_alg` ? Or to leave the functions in `rsa.h` alone and edit the lower level functions only ?

Reply:
For clarity, this is the documentaion/functionality I am referring to that would be needed to be changed in almost all `rsa.h` functions
```
 * \param hashlen  The length of the message digest.
 *                 This is only used if \p md_alg is #MBEDTLS_MD_NONE.
 ```

Reply:
@gilles-peskine-arm I started changing this, but it is not as quick as I hoped for. I would like to merge this fix rather sooner than later and going to submit a separate PR with the improvements you suggested.

Reply:
@nick-child-ibm Yes, I would like to change the RSA functions so that `hashlen` is checked rather than ignored.